src/coreclr/System.Private.CoreLib/src/System/Reflection/MdImport.cs

```diff
@@ -225,29 +231,69 @@ private bool Equals(MetadataImport import)
 
         #region Static Members
         [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern void _GetMarshalAs(IntPtr pNativeType, int cNativeType, out int unmanagedType, out int safeArraySubType, out string? safeArrayUserDefinedSubType,
-            out int arraySubType, out int sizeParamIndex, out int sizeConst, out string? marshalType, out string? marshalCookie,
+        private static extern unsafe Interop.BOOL GetMarshalAs(
+            IntPtr pNativeType,
+            int cNativeType,
+            out int unmanagedType,
+            out int safeArraySubType,
+            out sbyte* safeArrayUserDefinedSubType,
+            out int arraySubType,
+            out int sizeParamIndex,
+            out int sizeConst,
+            out sbyte* marshalType,
+            out sbyte* marshalCookie,
             out int iidParamIndex);
-        internal static void GetMarshalAs(ConstArray nativeType,
-            out UnmanagedType unmanagedType, out VarEnum safeArraySubType, out string? safeArrayUserDefinedSubType,
-            out UnmanagedType arraySubType, out int sizeParamIndex, out int sizeConst, out string? marshalType, out string? marshalCookie,
-            out int iidParamIndex)
-        {
 
-            _GetMarshalAs(nativeType.Signature, (int)nativeType.Length,
-                out int _unmanagedType, out int _safeArraySubType, out safeArrayUserDefinedSubType,
-                out int _arraySubType, out sizeParamIndex, out sizeConst, out marshalType, out marshalCookie,
-                out iidParamIndex);
-            unmanagedType = (UnmanagedType)_unmanagedType;
-            safeArraySubType = (VarEnum)_safeArraySubType;
-            arraySubType = (UnmanagedType)_arraySubType;
-        }
-        #endregion
+        internal static unsafe MarshalAsAttribute GetMarshalAs(ConstArray nativeType, RuntimeModule scope)
+        {
+            if (GetMarshalAs(
+                    nativeType.Signature,
+                    nativeType.Length,
+                    out int unmanagedTypeRaw,
+                    out int safeArraySubTypeRaw,
+                    out sbyte* safeArrayUserDefinedSubTypeRaw,
+                    out int arraySubTypeRaw,
+                    out int sizeParamIndex,
+                    out int sizeConst,
+                    out sbyte* marshalTypeRaw,
+                    out sbyte* marshalCookieRaw,
+                    out int iidParamIndex) != Interop.BOOL.TRUE)
+            {
+                throw new MetadataException(HResults.E_FAIL);
+            }
 
-        #region Internal Static Members
-        internal static void ThrowError(int hResult)
-        {
-            throw new MetadataException(hResult);
+            string? safeArrayUserDefinedTypeName = safeArrayUserDefinedSubTypeRaw == null ? null : new string(safeArrayUserDefinedSubTypeRaw);
```

Comment:
```suggestion
            string? safeArrayUserDefinedTypeName = safeArrayUserDefinedSubTypeRaw == null ? null : Encoding.UTF8.GetString(MemoryMarshal.CreateReadOnlySpanFromNullTerminated(safeArrayUserDefinedSubTypeRaw));
```
`new string(sbyte*)` uses current code page on Windows. This needs to be hardcoded to use UTF8.

`CreateReadOnlySpanFromNullTerminated` can be replaced by passing the string length returned by the metadata API on the native side. 

I do not have a strong preference about the exact pattern to use to create the managed string from UTF8 bytes.